Fixes #13228 - Update Core Deploy Documentation.

sbordet opened the pull request:
* Updated deploy documentation.
* Updated JNDI documentation.
* Added ext and eeN-ext module documentation.

sbordet committed:
Fixes #13228 - Update Core Deploy Documentation.

* Updated deploy documentation.
* Updated JNDI documentation.
* Added ext and eeN-ext module documentation.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

joakime commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -372,6 +379,10 @@ $ java -jar $JETTY_HOME/start.jar -Dwebapps.dir=/opt/myapps
 
 Properties specified in environment properties file (described in <<environment-configuration,this section>>) and in the web application properties file (described in <<web-application-files,this section>>) are available when processing the Jetty context XML file.
 
+Remember that Jetty Context XML files in the `$JETTY_BASE/webapps` directory can only refer to classes present in the environment class-path or module-path, and in the server class-path or module-path.
```
```suggestion
Remember that Jetty Context XML files in the `$JETTY_BASE/webapps` directory can only refer to classes present in the environment class-path (or module-path), and in the server class-path (or module-path).
```

joakime replied:
This section goes into far more detail than the similar (new) section below talking about com.acme.MyListenerFeature, perhaps we need a link (in both of these places) to new section on how classes are found (the environment classloader, the classpath, the modulepath, etc)

sbordet replied:
I added a link pointing to the class loader hierarchy section.

joakime commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -516,16 +511,16 @@ Jetty supports the following variants to be specified as virtual host names:
 A fully qualified domain name. It is important to list all variants as a site may receive traffic for both `www.hostname.com` and `hostname.com`.
 
-`*.hostname.com`::
+`+*.hostname.com+`::
```
What do the plus characters do?

```
 eg: `+ blah +`
```

sbordet replied:
The `+` disables inline formatting, so that e.g. the `*` is not interpreted as the start of bold, or `http://host` is not rendered as a link.

joakime reviewed (commented):
Looks good.

One sentence with "and", "or", and commas reads awkward.  An attempt at using parens to make it read better.

That same sentence is probably a bit more verbose than it needs to be.
A new sentence later is also referring to the same concepts, but isn't as verbose.
Perhaps a new section explaining the minutiae about our classloaders / classpath / modulepath is warranted (if it doesn't exist yet), and then linking to that new section from both of these 2 mentioned sentences.

gregw commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -22,17 +22,17 @@ The available environments are:
-* For Java EE 8, xref:modules/standard.adoc#eeN-deploy[`ee8-deploy`]
-* For Jakarta EE 9 up to Jakarta {ee-current-caps}, the corresponding xref:modules/standard.adoc#eeN-deploy[`ee<N>-deploy`]
+* For Jetty Static, the xref:modules/standard.adoc#static-deploy[`static-deploy`] module
+* For Jetty Core, the xref:modules/standard.adoc#core-deploy[`core-deploy`] module
```
You need to refer to "Core App" or "Core Application" rather than just "Core" everywhere in this document.    You should probably also use "Web Application" more in relation to EExx deployment, so we establish the naming better.

```suggestion
* For Jetty Core App, the xref:modules/standard.adoc#core-deploy[`core-deploy`] module
```

sbordet replied:
No, I am talking about the environments here, not the web applications.
I'll clarify that.

gregw commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -52,7 +52,7 @@ However, you can configure the `deployment-scanner` module to deploy web applica
 == Monitored Directories
 
-The `deployment-scanner` module scans the deploy directory specified by the `jetty.deploy.monitoredDir` property, by default `$JETTY_BASE/webapps`, for web applications to deploy.
+The `deployment-scanner` module scans the web applications deploy directory specified by the `jetty.deploy.webappsDir` property, by default `$JETTY_BASE/webapps`, for web applications to deploy.
```
```suggestion
The `deployment-scanner` module scans the web applications deploy directory specified by the `jetty.deploy.webappsDir` property, by default `$JETTY_BASE/webapps`, for web applications and core applications to deploy.
```

OR

```suggestion
The `deployment-scanner` module scans the web applications deploy directory specified by the `jetty.deploy.webappsDir` property, by default `$JETTY_BASE/webapps`, for contexts to deploy.
```

You might want to establish this language somewhere that a deployable context is a ContextHandler that could be a Webapp, coreapp, staticapp or directly configured core ContextHandler.

sbordet replied:
I rewrote the initial section, hopefully clearer now, and specified that "web application" refers to any kind -- static, core, javaee and jakartaee.

gregw commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -133,24 +133,31 @@ core-app
 
 For Jakarta EE web applications, the deployment unit is the usual `+*.war+` file or a directory with a `+*.war+` structure.
 
+Like for the <<web-application-format-core,Jetty Core Format>>, the `+*.war+` file or a directory can contain an optional in-application <<jetty-web-xml,Jetty Context XML file>> under the `WEB-INF` directory, that allows you to configure the web application deployment.
```
```suggestion
Like for the <<web-application-format-core,Jetty Core Application Format>>, the `+*.war+` file or a directory can contain an optional in-application <<jetty-web-xml,Jetty Context XML file>> under the `WEB-INF` directory, that allows you to configure the web application deployment.
```

gregw reviewed (changes requested):
This documentation uses "web application" to mean webapp or coreapp.   The coreapp is described as "core format" rather than "core app format".

I think that "context" should be used instead of "web application" and then make it clear that a "context" can be a: webapp, coreapp, staticapp, or otherwise configured core ContextHandler

sbordet commented:
@gregw "context" is a way too overloaded term, so I don't like it.

I can add a clarification that by "web application" I mean either static, core, or war.

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -13,26 +13,35 @@
+* <<web-application-format-jakarta,Java EE 8 web applications>> -- web applications written using the Servlet 4.0 (and associated specifications) in the `+javax.*+` packages.
+* <<web-application-format-jakarta,Jakarta EE 9 to Jakarta {ee-current-caps} web applications>> -- web applications written using the corresponding Servlet (and associated specifications) in the `+jakarta.*+` packages.
+
+In the following, the term "web application" is used to refer to either a Jetty Static web application, or Jetty Core web application, or Java EE 8 web application or Jakarta EE 9 to Jakarta {ee-current-caps} web application, unless specified otherwise.
```
When coming up with the coreapp vs webapp vs staticapp naming convention, we explicitly said that we would not use "web application" in the generic sense as it overloads that name in a confusing way and if somebody says "deploy the web application" we have no idea what they mean.

Let's pick a term for the generic.   It could just be "application", but I still don't mind "context".  Whilst "context" is somewhat overloaded, we do know that every "application" has a `ContextHandler` that is configured with a `contextPath` and that requests entering the "application" are given a `Context` instance that describes the "application" and that for web applications that context is a `ServletContext`.  It is all context!

janbartel replied:
I agree with @gregw . We specifically used the terms `coreapp`, `webapp` and `staticapp`. We cannot not now refer to all of them as web applications without nullifying the clear delineation of the types of application. So IMHO they should be referred to as a group as `applications`, which is a widely understood term, and fits nicely with the specialized cases of `core application`, `web application` and `static application`.

gregw reviewed (changes requested):
Sorry, but I'm a big -1 on using "web application" both in the generic and specific sense.  It is too confusing and will render the term meaningless as it will always require further qualification.  It has to be one or the other, not both.

See my comment for further discussion.

sbordet commented:
@gregw let's discuss this. I am not opposed to just use "application", but there are so many other places where we use "web application", for example the "webapps deploy directory" is called "$JETTY_BASE/webapps", not just "apps", same with property `jetty.deploy.webappsDir`, `DeploymentScanner.addWebappsDirectory()`, the client documentation refers to "making a request to a web application", and colloquially it is said to "make a request to that web application" without knowing how exactly it is packaged.

Aren't we trying to give "web application" a more specific meaning that it does not have, certainly not for non-Jetty committers?

Had we named "JakartaWarContext" or "WarContext" what we now call "oej.ee11.webapp.WebAppContext", would you have a problem using "web application"?

I still don't see what makes a static web application different from a Jakarta web application.
From the end user (browser) point of view, there would be no way to tell the difference.
From the developer point of view, it is the unit of packaging that needs to be deployed, and they get deployed in the same directory, so they walk the same, quack the same, and look the same duck to me.

Anyway, I tried to make the documentation as consistent as possible, so it would be trivial to replace "web application" with just "application" in the narrative.

janbartel commented on documentation/jetty/modules/operations-guide/pages/jndi/index.adoc:
```diff
@@ -13,71 +13,69 @@
-* For Java EE 8, `ee8-plus`
-* For Java EE 9, `ee9-plus`
-* For Java {ee-current-caps}, `{ee-current}-plus`
+* For Static and Core web applications, the `plus` module.
```
```suggestion
* For Static and Core applications, the `plus` module.
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/jndi/index.adoc:
```diff
@@ -13,71 +13,69 @@
-* For Java EE 9, `ee9-plus`
-* For Java {ee-current-caps}, `{ee-current}-plus`
+* For Static and Core web applications, the `plus` module.
+* For Jakarta {ee-all-caps} the correspondent `{ee-all}-plus` module.
```
```suggestion
* For Jakarta {ee-all-caps} web applications the correspondent `{ee-all}-plus` module.
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/jndi/index.adoc:
```diff
@@ -13,71 +13,69 @@
+If your JNDI resources need extra libraries, for example database driver libraries, that are not located inside your web application, then you must ensure they are on environment class-path.
 
-You can now declare JNDI resources and reference them within your webapps.
+You can enable the xref:modules/standard.adoc#ext[`ext`] module for Core web applications, or the specific xref:modules/standard.adoc#eeN-ext[`{ee-all}-ext`] module for Jakarta web applications, and place your extra libraries respectively in `$JETTY_BASE/lib/ext` or `$JETTY_BASE/lib/{ee-all}/ext` to make them available in the environment class-path.
```
```suggestion
You can enable the xref:modules/standard.adoc#ext[`ext`] module for Core applications, or the specific xref:modules/standard.adoc#eeN-ext[`{ee-all}-ext`] module for Jakarta web applications, and place your extra libraries respectively in `$JETTY_BASE/lib/ext` or `$JETTY_BASE/lib/{ee-all}/ext` to make them available in the environment class-path.
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/jndi/index.adoc:
```diff
@@ -13,71 +13,69 @@
+You can enable the xref:modules/standard.adoc#ext[`ext`] module for Core web applications, or the specific xref:modules/standard.adoc#eeN-ext[`{ee-all}-ext`] module for Jakarta web applications, and place your extra libraries respectively in `$JETTY_BASE/lib/ext` or `$JETTY_BASE/lib/{ee-all}/ext` to make them available in the environment class-path.
 
-== Declaring resources
+You can now define JNDI resources and reference them within your web applications.
```
```suggestion
You can now define JNDI resources and reference them within your applications.
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/jndi/index.adoc:
```diff
@@ -13,71 +13,69 @@
-You make these declarations in Jetty XML files that are either _external_  or _internal_ to your webapp.
-A server or context XML file is external to your webapp.
-The special `WEB-INF/jetty-{ee-all}-env.xml` file is internal to your webapp. 
+JNDI resources are referenced by web application via JNDI names.
```
```suggestion
JNDI resources are referenced by applications via JNDI names.
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/jndi/index.adoc:
```diff
@@ -13,71 +13,69 @@
-A server or context XML file is external to your webapp.
-The special `WEB-INF/jetty-{ee-all}-env.xml` file is internal to your webapp. 
+JNDI resources are referenced by web application via JNDI names.
+For Jakarta web applications, the JNDI names must be declared via `env-entry`, `resource-ref` and `resource-env-refs` in `web.xml`, `web-fragment.xml` or Jetty's `override-web.xml`.
```
I thin k this is confusing putting it this way around. I preferred the original text that describes for Jakarta web applications, you need to declare a resource in a Jetty XML file that is either  _internal_ or _external_  to your webapp. You then reference the names of these resources in `web.xml`, `web-fragment.xml` or `web-override.xml`.

janbartel commented on documentation/jetty/modules/operations-guide/pages/jndi/index.adoc:
```diff
@@ -13,71 +13,69 @@
 <<tx,`org.eclipse.jetty.plus.jndi.Transaction`>>::
-For a JTA manager
+<<tx,`org.eclipse.jetty.{ee-all}.plus.jndi.Transaction`>>::
+Used for a transaction resource.
```
Prefer original text. It is a `JTA` transaction manager.

janbartel commented on documentation/jetty/modules/operations-guide/pages/jndi/index.adoc:
```diff
@@ -143,7 +140,7 @@ This example configures a http://db.apache.org/derby[Derby] DataSource named `jd
 ----
 
-This would be linked into the webapp's JNDI namespace via an entry in a `web.xml` like so:
+This would be linked into the web application JNDI namespace via an entry in a `web.xml` like so:
```
```suggestion
This would be linked into the web application's JNDI namespace via an entry in a `web.xml` like so:
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/jndi/index.adoc:
```diff
@@ -189,9 +184,8 @@ Here is an example of binding an http://activemq.apache.org[ActiveMQ] in-JVM con
 ----
 
-The corresponding entry in `web.xml` to bind the ConnectionFactory into the webapp's JNDI namespace would be:
+The corresponding entry in `web.xml` to bind the ConnectionFactory into the web application JNDI namespace would be:
```
```suggestion
The corresponding entry in `web.xml` to bind the ConnectionFactory into the web application's JNDI namespace would be:
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/modules/standard.adoc:
```diff
@@ -219,7 +219,7 @@ include::{jetty-home}/modules/deployment-scanner.mod[tags=documentation]
 Among the configurable properties, the most relevant are:
 
-`jetty.deploy.monitoredDir`::
+`jetty.deploy.webappsDir`::
```
This really should be changed to the `jetty.deploy.appsDir`.

sbordet replied:
I don't think it is a good idea.
We certainly don't want to change the name of `$JETTY_BASE/webapps` into `$JETTY_BASE/apps`.

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -13,26 +13,35 @@
-* The Java EE 8 Environment -- Supports Servlet 4.0 (and associated specifications) in the `+javax.*+` packages.
-* Any version from the Jakarta EE 9 Environment up to the Jakarta {ee-current-caps} Environment -- Supports corresponding Servlet (and associated specifications) in the `+jakarta.*+` packages.
+* The `static` environment, for Jetty Static web applications.
+* The `core` environment, for Jetty Core web applications.
```
```suggestion
* The `core` environment, for Jetty Core applications.
```

gregw replied:
```suggestion
* The `core` environment, for Jetty Core web applications and XML configured `ContextHandler`s.
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -13,26 +13,35 @@
-* The Jetty Core Environment -- Supports web applications written against the Jetty `Handler` APIs, without any Servlet dependencies.
-* The Java EE 8 Environment -- Supports Servlet 4.0 (and associated specifications) in the `+javax.*+` packages.
-* Any version from the Jakarta EE 9 Environment up to the Jakarta {ee-current-caps} Environment -- Supports corresponding Servlet (and associated specifications) in the `+jakarta.*+` packages.
+* The `static` environment, for Jetty Static web applications.
```
```suggestion
* The `static` environment, for Jetty Static applications.
```

gregw replied:
We have agreed on "web application", so I'll resolve for clarity.

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -13,26 +13,35 @@
+* The `{ee-all}` environments, for Jakarta EE web applications.
 
-This means that you can simultaneously deploy an old Java EE 8 web application, say `old-ee8.war`, alongside a new Jakarta {ee-current-caps} web application, say `new-{ee-current}.war`, alongside a web application that only uses the core Jetty `Handler` APIs, all in the same Jetty server.
+You can simultaneously deploy an old Java EE 8 web application, say `old-ee8.war`, alongside a new Jakarta {ee-current-caps} web application, say `new-{ee-current}.war`, alongside a web application that only uses the Jetty Core `Handler` APIs, say `backoffice-core.xml`, all in the same Jetty server.
```
```suggestion
You can simultaneously deploy an old Java EE 8 web application, say `old-ee8.war`, alongside a new Jakarta {ee-current-caps} web application, say `new-{ee-current}.war`, alongside an application that only uses the Jetty Core `Handler` APIs, say `backoffice-core.xml`, all in the same Jetty server.
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -52,7 +61,7 @@ However, you can configure the `deployment-scanner` module to deploy web applica
 == Monitored Directories
 
-The `deployment-scanner` module scans the deploy directory specified by the `jetty.deploy.monitoredDir` property, by default `$JETTY_BASE/webapps`, for web applications to deploy.
+The `deployment-scanner` module scans the web applications deploy directory specified by the `jetty.deploy.webappsDir` property, by default `$JETTY_BASE/webapps`, for web applications to deploy.
```
```suggestion
The `deployment-scanner` module scans the applications deploy directory specified by the `jetty.deploy.appsDir` property, by default `$JETTY_BASE/webapps`, for applications to deploy.
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -87,7 +96,7 @@ The web application base resource directory can be outside of `$JETTY_BASE/webap
 
 [[web-application-format-static]]
-=== Jetty Static Format
+=== Jetty Static Web Application Format
```
```suggestion
=== Jetty Static Application Format
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -118,7 +127,7 @@ jetty.deploy.baseResource.dirAllowed=false
 
 [[web-application-format-core]]
-=== Jetty Core Format
+=== Jetty Core Web Application Format
```
```suggestion
=== Jetty Core Application Format
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -118,7 +127,7 @@ jetty.deploy.baseResource.dirAllowed=false
-=== Jetty Core Format
+=== Jetty Core Web Application Format
 
 For Jetty Core web applications, the web application format is a directory with the following structure:
```
```suggestion
For Jetty Core applications, the web application format is a directory with the following structure:
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -133,24 +142,31 @@ core-app
-    └── acme-util.jar
+├── lib            # Where web application Java libraries reside.
+│   └── acme-util.jar
+└── jetty-web.xml  # Web application specific configuration.
```
This file is now badly named. It is nothing to do with `web.xml` but rather another Jetty XML context XML file that is processed after `web.xml`.

janbartel replied:
Also, how are you arranging for this to be processed for a core application, where there is no concept of `Configuration`s  that would read such a file?

sbordet replied:
@janbartel this is done by `CoreAppContext` directly.

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -133,24 +142,31 @@ core-app
+Since static files can only be served from the `static/` directory, the `classes/` and `lib/` directories are implicitly hidden, as well as any other directory such as `includes/` or `fragments/`, and any file in the directory root.
 
-This implies that the `classes/` and `lib/` directories are automatically hidden, as well as any other directory such as `includes/` or `fragments/`.
+The `classes/` directory stores the web application's `+*.class+` files, while the `lib/` directory stores the web application's libraries, typically `+*.jar+` files.
```
```suggestion
The `classes/` directory stores the application's `+*.class+` files, while the `lib/` directory stores the application's libraries, typically `+*.jar+` files.
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -133,24 +142,31 @@ core-app
 
 A notable difference with `+*.war+` files is that static files are only served from the `static/` directory (with the `static/` prefix removed).
 This means that files in the directory root, or under any other directory that is not `static/`, are hidden and if requested they will result in a `404 Not Found` response.
+If the `static/` directory is missing, then no static files can be served, and therefore all requests are processed dynamically by the ``Handler``s configured in the web application.
```
```suggestion
If the `static/` directory is missing, then no static files can be served, and therefore all requests are processed dynamically by the ``Handler``s configured in the application.
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -133,24 +142,31 @@ core-app
+Both these directories are optional.
 
-The `classes/` directory stores the web application's `+*.class+` files, while the `lib/` directory stores the web application's libraries.
+The <<jetty-web-xml,`jetty-web.xml`>> file is an optional in-application Jetty Context XML file, that allows you to configure the web application deployment.
```
```suggestion
The <<jetty-web-xml,`jetty-web.xml`>> file is an optional in-application Jetty Context XML file, that allows you to configure the application deployment.
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -372,6 +388,11 @@ $ java -jar $JETTY_HOME/start.jar -Dwebapps.dir=/opt/myapps
 
 Properties specified in environment properties file (described in <<environment-configuration,this section>>) and in the web application properties file (described in <<web-application-files,this section>>) are available when processing the Jetty context XML file.
 
+Remember that Jetty Context XML files in the `$JETTY_BASE/webapps` directory can only refer to classes present in the environment class-path/module-path, and in the server class-path/module-path.
```
Also the jvm's classpath

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -427,24 +452,37 @@ $JETTY_BASE
+* To perform configuration of the web application that can only occur _after_ the web application's class-path has been created, for example because it references classes present in web application itself.
+
+[[jetty-web-xml-core]]
+=== In-Application Jetty Context XML File for Core Web Application Deployments
```
```suggestion
=== In-Application Jetty Context XML File for Core Application Deployments
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -427,24 +452,37 @@ $JETTY_BASE
 
 This file can be useful:
 
 * To place all configuration inside your web application archive.
```
```suggestion
* To place all configuration inside your application archive.
```

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -427,24 +452,37 @@ $JETTY_BASE
+[[jetty-web-xml-core]]
+=== In-Application Jetty Context XML File for Core Web Application Deployments
+
+The file must be in the root of the web application and named `jetty-web.xml`, see <<web-application-format-core,this section>> for an example.
```
See previous comment - how is this being processed?

janbartel replied:
I see now that the `CoreContextHandler` is specifically looking for a file called `jetty-web.xml` and processing it. I don't like this - I know we have traditionally had a file called `jetty-web.xml` in ee web apps, but it has always been a misnomer - it appears like it's a jetty-specific extension to `web.xml` but really it's just a jetty context XML file. In addition, the expectation is that a `jetty-{ee}-web.xml` will be configuring the `WebAppContext`, not the `Server`, nor any other object (that's why the `ee` is in the name now!).  So I think the name is inappropriate in general, and very inappropriate here. The name should reflect that it is a context XML file, and that is configuring the `ContextHandler`, and ideally that it is the last configuration step before the app is started. I could live with something like `jetty-context.xml` for core apps.

joakime replied:
The `jetty-web.xml` follows the same rules as the WebAppContext.
They configure the Context, that's it, nothing else.
The scare tactic that it can configure the Server isn't true.

https://github.com/jetty/jetty.project/blob/jetty-12.1.x/jetty-core/jetty-coreapp/src/main/java/org/eclipse/jetty/coreapp/CoreAppContext.java#L427

https://github.com/jetty/jetty.project/blob/434d1d934c9bf22c121035cce3e3302fd5fba5d6/jetty-core/jetty-coreapp/src/main/java/org/eclipse/jetty/coreapp/CoreAppContext.java#L427

This is enforced with the `XmlConfiguration.configure(this)` where the input parameter instance is the only thing that can be configured, if you specify any other class in the XML that isn't that class, it will fail.

joakime replied:
> So I think the name is inappropriate in general, and very inappropriate here. The name should reflect that it is a context XML file, and that is configuring the `ContextHandler`, and ideally that it is the last configuration step before the app is started. I could live with something like `jetty-context.xml` for core apps.

That ship has sailed a long time ago.

Having a different name for a file in CoreAppContext that serves the same role / purpose / featureset that WebAppContext does is an even worse suggestion.

If you want a new / different name, then we need to fix it everywhere, even in WebAppContext.
But know that we cannot eliminate `jetty-web.xml` in WebAppContext, pretty much ever, as there's too many wars already built with that included now.

joakime replied:
To me, `WEB-INF/web.xml` is a "Jakarta Web App Descriptor".
And `jetty-web.xml` is a "Jetty XML for Web Contexts".

There is no conflict.

janbartel replied:
Just because it's wrong in web apps, doesn't mean we have to perpetuate that in the brand new concept of core apps. As this is a green field, we have freedom to call the file by whatever name makes the most sense. As it has always and only been used to configure a `WebAppContext`, it makes no sense to call it `jetty-web.xml`.

janbartel commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -427,24 +452,37 @@ $JETTY_BASE
 
 * To place all configuration inside your web application archive.
-* To perform configuration of the web application that can only occur _after_ the web application's class-path has been created, for example because it references classes present in `WEB-INF/classes` or `WEB-INF/lib`.
+* To perform configuration of the web application that can only occur _after_ the web application's class-path has been created, for example because it references classes present in web application itself.
```
```suggestion
* To perform configuration of the web application that can only occur _after_ the application's class-path has been created, for example because it references classes present in the application itself.
```

joakime commented:
Just adding a statement about terminology in today's terms ...

![app](https://github.com/user-attachments/assets/25f42bbe-e372-4335-85f6-140d6b200cb3)

janbartel commented:
@sbordet I am unconvinced that the generic name for core apps, static apps and web apps is `"web applications"`. In some sense, as they are all deployed on a web server, it's tautological to call them `"web applications"`. Moreover, I think it dilutes the benefit of our clear nomenclature: core apps, static apps, web apps. There is also just as much work - if not more - in going back over the documentation to change references from `"web applications"` to `"ee web applications"` as there is in simply referring to `"applications"` when meaning any of the 3 types.

sbordet committed:
* Contextualized the use of "web application" when it explicitly refers to Jakarta EE and/or Jetty.
* Fixed WebSocket documentation.
* Fixed Quickstart documentation.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Colloquial "webapp" converted to "web application".

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@janbartel I think "web applications" is the usual colloquial term that everyone uses.

In particular, being so strict in the interpretation of the term is potentially confusing: if "web applications" (intended as Jakarta EE ones only) are deployed from `$JETTY_BASE/webapps`, where would a "core app" be deployed from?

I have reviewed the documentation and changed references to "web application" to "Jakarta EE web application" only where it was necessary, and not surprisingly there were not so many places to clarify.
Most of the documentation is valid for all types of web applications.
The opposite (changing everything to just "application") would be a much, much larger work.

While reviewing the documentation, it was very clear that "web application" was used colloquially without any specific reference to Jakarta EE, so I think we should retain that colloquial meaning.

joakime commented:
@janbartel you are seeing this entirely from your deep background developing the code of Jetty, not from the users point of view.

Lets take a look at a few descriptions.

https://www.britannica.com/topic/Web-application

> Web application, computer program stored on a remote server and run by its users via a Web browser.

https://en.wikipedia.org/wiki/Web_application

> A web application (or web app) is application software that is created with web technologies and runs via a web browser.[1][2] Web applications emerged during the late 1990s and allowed for the server to dynamically build a response to the request, in contrast to static web pages.[3]

With those two descriptions, Jakarta Web App, Core App, and even Static App are "web apps".

Yes, even our Static App fits, as it's based on ResourceHandler, which doesn't _just_ serve static web pages, but also dynamic serving of welcome files, compression, content-types, file listings, etc.

Heck, even our DefaultHandler that serves an error page pointing at the deployed contexts is a web app.

gregw commented:
I can live with "Web Application".   If we went to "Application" for the generic, then we'd be an "Application Server" and that implies full EE support

gregw commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -13,26 +13,35 @@
 
 = Web Application Deployment
 
+The Jetty server supports the deployment of these types of web applications:
+
+* <<web-application-format-static,Jetty Static web applications>> -- web applications made of static files only.
+* <<web-application-format-core,Jetty Core web applications>> -- web applications written against the Jetty `Handler` APIs, without any Servlet dependencies.
+* <<web-application-format-jakarta,Java EE 8 web applications>> -- web applications written using the Servlet 4.0 (and associated specifications) in the `+javax.*+` packages.
+* <<web-application-format-jakarta,Jakarta EE 9 to Jakarta {ee-current-caps} web applications>> -- web applications written using the corresponding Servlet (and associated specifications) in the `+jakarta.*+` packages.
+
+In the following, the term "web application" is used to refer to either a Jetty Static web application, or Jetty Core web application, or Java EE 8 web application or Jakarta EE 9 to Jakarta {ee-current-caps} web application, unless specified otherwise.
```
We can also deploy directly configured `ContextHandlers` with a context XML file:

```suggestion
The Jetty server supports the deployment of the following type of web applications:

* <<web-application-format-jakarta,Jakarta EE 9 to Jakarta {ee-current-caps} web applications>> -- web applications written using the corresponding Servlet (and associated specifications) in the `+jakarta.*+` packages.
* <<web-application-format-jakarta,Java EE 8 web applications>> -- web applications written using the Servlet 4.0 (and associated specifications) in the `+javax.*+` packages.
* <<web-application-format-core,Jetty Core web applications>> -- web applications written against the Jetty `Handler` APIs in a Jetty specific deployment structure, without any Servlet dependencies.
* <<web-application-format-static,Jetty Static web applications>> -- web applications made of static files only.
* <<web-application-context-handler,Jetty Context Handler>> -- web application made by an XML configured `ContextHandler` instance.

In the following, the term "web application" is used to refer to either a Jetty Static web application, or Jetty Core web application, or directly configured `ContextHandler`, or Java EE 8 web application or Jakarta EE 9 to Jakarta {ee-current-caps} web application, unless specified otherwise.
```

janbartel replied:
@joakime if we use "web application" generically, then everywhere where we have documentation that is specific to "ee web applications" we need to edit it and fix.

sbordet replied:
@janbartel I have done that.

Of course, I might have missed a few, feel free to point them out.

gregw reviewed (changes requested):
I'm on board with the approach of using "Web Application" generically and being more specific only when necessary.    However the current document has missed an important variation - an XML configured `ContextHandler` from a context XML file is a valid deployment of a webapplication that is not a Jetty Core Web Application, but is a Web Application using core APIs !

jmcc0nn3ll commented:
I think it might behoove us to start an asciidoc [description list ](https://docs.asciidoctor.org/asciidoc/latest/lists/description/) page (or something similar) where we put a formal description of what these given terms mean in the context of Jetty. Then we can link the word 'web application' to that description when it makes sense or may be confusing. That way we are not strictly relying on the readers contextual understanding so much, especially when we are using largely bespoke terms like Core App which means using internal apis.

sbordet committed:
Updates from review.

Added section on Jetty `ContextHandler` web applications.
Documented `MovedContextHandler`.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'fix/jetty-12.1.x/13228/update-deploy-docs'.

gregw commented:
> I think it might behoove us to start an asciidoc [description list ](https://docs.asciidoctor.org/asciidoc/latest/lists/description/) page (or something similar) where we put a formal description of what these given terms mean in the context of Jetty. Then we can link the word 'web application' to that description when it makes sense or may be confusing. That way we are not strictly relying on the readers contextual understanding so much, especially when we are using largely bespoke terms like Core App which means using internal apis.

Let's just add a Glossary section to the document!

sbordet commented:
> Let's just add a Glossary section to the document!

Nobody reads glossaries!
Not in these times of Google and AI.

gregw commented:
> > Let's just add a Glossary section to the document!
> 
> Nobody reads glossaries! Not in these times of Google and AI.

Glossaries train google and AI. Plus it would be as much as for us as users. Somewhere to record consistent language usage. Better than in a file that nobody sees.

joakime commented:
> > Let's just add a Glossary section to the document!
> 
> Nobody reads glossaries! Not in these times of Google and AI.

Just because you don't read them doesn't mean others don't read them too.

Heck, look at how many glossary's there are at google themselves!

https://www.google.com/search?q=glossary+site%3Agoogle.com

joakime commented:
Heck, the number of times I've flipped to the back of a Sci Fi book to read it's glossary to understand a new term is countless. (Just wish all scifi and fantasy books did it)

gregw reviewed (approved):
This is a good improvement, but we definitely need a glossary. We do define the difference between EE, Core and `ContextHandler` web application in the deployer doc, but having a glossary to look that up would be a good way to help those that use them and to train AI.

janbartel commented on documentation/jetty/modules/operations-guide/pages/jsf-taglibs/index.adoc:
```diff
@@ -13,7 +13,7 @@
 = JavaServer Faces TagLibs
 
-If you want to use JSF with your webapp, you should copy the relevant jars from your implementation of choice into your `$JETTY_BASE` directory, ideally into `$JETTY_BASE/lib/ext`.
+If you want to use JSF with your web application, you should copy the relevant jars from your implementation of choice into your `$JETTY_BASE` directory, ideally into `$JETTY_BASE/lib/ext`.
```
This is exactly the point I've been trying to make! This change is incorrect: as you want to use `web application` as a generic term, this sentence implies that you can use JSF with a core app or a static app, which is emphatically _not_ the case. To ensure the consistency of the nomenclature, we need to change every occurrence of `web application` where it _only_ applies to a Servlet Spec type application to read `ee web application`.  It would be much much simpler to just use `application` in the _new_  documentation of core and static apps. If there's no appetite to do that, then the documentation needs to be updated to read `ee web application` where applicable.

sbordet replied:
I already did a good pass about specifying Jakarta EE web application.

I fixed the one few lines below, but missed this one.

It's obvious that the documentation is referring to a Jakarta EE technology that you can only use in a Jakarta EE web application.

Oh, look, the title is wrong, the reference to the `ext` module is wrong, the javadoc link is wrong, the EE version is hardcoded to `ee9`.

I'll fix everything, including "web application" 🙄

sbordet committed:
Updated the Jakarta Faces documentation after review comments.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

janbartel reviewed (approved):
Well, I still don't like the web application terminology, but I'm in the minority, so I'm giving this a +1.
